### DISCUSSION TITLE
How to run a user-defined function in .cue file without modifying the cue source code?

### DISCUSSION THREAD

**Discussion by ningan123** [cue-lang/cue#1685] [Q&A] [answered]
I want to define a variable in the .cue file, the variable's value is an available port in a real host. To ensure the port is an available port, I want to run a user-defined function in the .cue file.

```cue
// test.cue

import "xxx"
availablePortsList : GetAvailablePort()
```

```go
// test.go

func GetAvailablePort() []int{
   ...
   availablePorts := ...
   ...
   return availablePorts
}


```

```go
// main.go

import (
	"fmt"
	"cuelang.org/go/cue/ast"
	"cuelang.org/go/cue/load"
	"cuelang.org/go/cue/cuecontext"
)

func main(){
    ctx := cuecontext.New() 
    entryPoints := []string{"test.cue"}
    b := ctx.BuildInstance(...)
    ...
    fmt.Printf("b: %s\n", b)
}
```

When I run
```
go run test.go
```

I expect to see:
```
availablePortsList: [1111, 1112, 1113, ...]
```

1. I don't want to change the .cue source code
2.  All I need to do is run the main.go

With this, how can I do it? Thank u.

**Comment by myitcv** [comment]
@ningan123 I've converted your issue to a discussion because we use issues for tracking bugs and feature requests. Thanks!

**Comment by myitcv** [comment, marked as answer]
> I want to run a user-defined function in the .cue file.

Please see https://github.com/cue-lang/cue/issues/350. This is not supported within CUE evaluation because it would severely fragment the CUE ecosystem. 

Given that you are looking to load and evaluate CUE from the Go API, you could do something like the following:

```
go mod tidy
go run main.go

-- cue.mod/module.cue --
module: "mod.com"
-- go.mod --
module mod.com

go 1.18

require cuelang.org/go v0.4.3

-- main.go --
package main

import (
	"fmt"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/load"
)

func main() {
	ctx := cuecontext.New()
	bps := load.Instances([]string{"."}, nil)
	v := ctx.BuildInstance(bps[0])
	v = v.FillPath(cue.ParsePath("availablePortsList"), GetAvailablePort())
	fmt.Printf("v: %v\n", v)
}

func GetAvailablePort() []int {
	return []int{1111, 1112, 1113}
}
-- x.cue --
package x

availablePortsList: [...int]
```

which gives:

```
> go mod tidy
> go run main.go
[stdout]
v: {
        availablePortsList: [1111, 1112, 1113]
}
```

**Reply by verdverm** [reply]
@myitcv should that Go directive be 1.18 rather than 1.19?

**Reply by myitcv** [reply]
> @myitcv should that Go directive be 1.18 rather than 1.19?

Thanks. I run tip locally - probably need to default to 1.18.
